### DISCUSSION TITLE
C# Issues

### DISCUSSION THREAD

**Discussion by jamesdimick** [godot-jolt/godot-jolt#442] [Questions] [answered]
So, because of the current [issues](https://github.com/godot-jolt/godot-jolt/issues/270) in Godot, we are not able to use any of the physics functions from C# code. But there is the patch [here](https://github.com/godotengine/godot/pull/75955) that works around it. So, would it work then to try running a version of Godot compiled with that patch applied? _(Would have to figure out how to do that though since I haven't tried it before)_

It just seems like there isn't a lot of progress on the Godot side yet, and I'd really like to be able to test out Godot Jolt.

**Comment by mihe** [comment, marked as answer]
> So, would it work then to try running a version of Godot compiled with that patch applied?

Yes, but note that the PR is technically applied to Godot's `master` branch, which targets a newer version of GDExtension that is not currently compatible with Godot Jolt.

Thankfully though, it seems like not much has happened in the particular files that the PR touches, so the change can (as of right now) be applied to the `4.0.3-stable` tag of Godot as well.

I'm assuming you want to test this on Linux, considering your previous issue here, in which case it would go something like this:

(This all assumes that you have the .NET SDK already installed, as well as all [the other development packages](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_linuxbsd.html#distro-specific-one-liners) that Godot requires.)

```sh
# Clone the Godot repository
git clone https://github.com/godotengine/godot.git

# Navigate into the new directory
cd godot

# Check out the commit for 4.0.3-stable
git checkout tags/4.0.3-stable

# Download patch generated for the pull request
wget https://github.com/godotengine/godot/pull/75955.patch

# Apply patch
git apply ./75955.patch

# Build editor
scons p=linuxbsd target=editor module_mono_enabled=yes production=yes

# Build the debug export template (you can skip this or save it for later)
scons p=linuxbsd target=template_debug module_mono_enabled=yes production=yes

# Build the release export template (you can skip this or save it for later)
scons p=linuxbsd target=template_release module_mono_enabled=yes production=yes

# Generate .NET glue source code
./bin/godot.linuxbsd.editor.x86_64.mono --headless --generate-mono-glue modules/mono/glue

# Build the .NET assemblies
./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform=linuxbsd
```

You might see errors about missing the .NET SDK for some of the later commands there, but if you're able to push past them and still see a successful completion, you should be good.

**Reply by jamesdimick** [reply]
> > So, would it work then to try running a version of Godot compiled with that patch applied?
> 
> Yes, but note that the PR is technically applied to Godot's `master` branch, which targets a newer version of GDExtension that is not currently compatible with Godot Jolt.
> 
> Thankfully though, it seems like not much has happened in the particular files that the PR touches, so the change can (as of right now) be applied to the `4.0.3-stable` tag of Godot as well.
> 
> I'm assuming you want to test this on Linux, considering your previous issue here, in which case it would go something like this:
> 
> (This all assumes that you have the .NET SDK already installed, as well as all [the other development packages](https://docs.godotengine.org/en/stable/contributing/development/compiling/compiling_for_linuxbsd.html#distro-specific-one-liners) that Godot requires.)
> 
> ```shell
> # Clone the Godot repository
> git clone https://github.com/godotengine/godot.git
> 
> # Navigate into the new directory
> cd godot
> 
> # Check out the commit for 4.0.3-stable
> git checkout tags/4.0.3-stable
> 
> # Download patch generated for the pull request
> wget https://github.com/godotengine/godot/pull/75955.patch
> 
> # Apply patch
> git apply ./75955.patch
> 
> # Build editor
> scons p=linuxbsd target=editor module_mono_enabled=yes production=yes
> 
> # Build the debug export template (you can skip this or save it for later)
> scons p=linuxbsd target=template_debug module_mono_enabled=yes production=yes
> 
> # Build the release export template (you can skip this or save it for later)
> scons p=linuxbsd target=template_release module_mono_enabled=yes production=yes
> 
> # Generate .NET glue source code
> ./bin/godot.linuxbsd.editor.x86_64.mono --headless --generate-mono-glue modules/mono/glue
> 
> # Build the .NET assemblies
> ./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform=linuxbsd
> ```
> 
> You might see errors about missing the .NET SDK for some of the later commands there, but if you're able to push past them and still see a successful completion, you should be good.

So, that works! Awesome, thanks! :)

--

The only issue I'm having now is some sort of positional snapping going on with my test third-person physics-based character. I didn't have that issue with Godot Physics, so I'm not sure what it could be. My code basically boils down to this simplified example:

```
public override void _IntegrateForces( PhysicsDirectBodyState3D state )
{
	float inputV = Input.GetAxis( "gameplay_backward", "gameplay_forward" );
	float inputH = Input.GetAxis( "gameplay_right", "gameplay_left" );

	Camera3D cam = GetNode<Camera3D>( "Camera" );
	Vector3 camFwd = new Vector3( -cam.GlobalTransform.Basis.Z.X, 0f, -cam.GlobalTransform.Basis.Z.Z );
	Vector3 camRght = new Vector3( -cam.GlobalTransform.Basis.X.X, 0f, -cam.GlobalTransform.Basis.X.Z );

	Vector3 direction = ( camFwd.Normalized() * inputV ) + ( camRght.Normalized() * inputH );
	direction = direction.Normalized();

	Vector3 newVelocity = direction * 50f;
	newVelocity.Y = 0f;

	state.ApplyCentralForce( newVelocity );
}
```

Basically, when moving the player with inputs it seems to work. But then when I let off the inputs and the player is coming to a stop, there is this weird snapping behavior _(or something)_ where the player moves over a little bit, even though I'm no longer applying inputs at that point.

I would think it could be an issue with my code, but the fact is that same code, same scene, same setup, etc worked just fine in Godot Physics, and even Bullet Physics back in Godot 3, so I'm not so sure.

**Reply by jamesdimick** [reply]
Oh and I know that that code isn't very good for a player controller. It is just the basics that are left after I comment out/remove all other code trying to narrow down the issue. If that basic movement can't work, then more advanced stuff wouldn't work either I guess.

**Reply by mihe** [reply]
Would you mind opening an issue for it, with a repro project attached? I'm not exactly sure what your scene setup is in that example. It's fine if it's in C#.

**Reply by jamesdimick** [reply]
> Would you mind opening an issue for it, with a repro project attached? I'm not exactly sure what your scene setup is in that example. It's fine if it's in C#.

Ok, done. :)
